File: Windows_and_Linux/history_manager.py
"""
History Manager
---------------

Singleton service for recording, persisting, and managing command execution history.
Stores entries in command_history.json alongside config.json.
"""

import json
import logging
import os
import sys
from typing import List, Optional

from history_entry import HistoryEntry

MAX_HISTORY_ENTRIES = 100


class HistoryManager:
    """Manages command execution history with JSON file persistence."""

    _instance: Optional["HistoryManager"] = None

    def __init__(self):
        self.entries: List[HistoryEntry] = []
        self._file_path = os.path.join(os.path.dirname(sys.argv[0]), "command_history.json")
        self.load()

    @classmethod
    def shared(cls) -> "HistoryManager":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    # ── Persistence ──────────────────────────────────────────────

    def load(self):
        if os.path.exists(self._file_path):
            try:
                with open(self._file_path, "r", encoding="utf-8") as f:
                    data = json.load(f)
                self.entries = [HistoryEntry.from_dict(d) for d in data]
                logging.debug(f"Loaded {len(self.entries)} history entries")
            except Exception as e:
                logging.error(f"Failed to load history: {e}")
                self.entries = []
        else:
            self.entries = []

    def save(self):
        try:
            with open(self._file_path, "w", encoding="utf-8") as f:
                json.dump([e.to_dict() for e in self.entries], f, indent=2)
        except Exception as e:
            logging.error(f"Failed to save history: {e}")

    # ── Recording ────────────────────────────────────────────────

    def record(
        self,
        command_name: str,
        input_text: str,
        output_text: str,
        model_name: str = "",
        source_app_name: str = "",
        matched_profile_name: str = "",
    ):
        """Record a new history entry. Respects the max-entries cap."""
        entry = HistoryEntry(
            command_name=command_name,
            input_text=input_text,
            output_text=output_text,
            model_name=model_name,
            source_app_name=source_app_name,
            matched_profile_name=matched_profile_name,
        )
        self.entries.insert(0, entry)

        # Prune to cap
        if len(self.entries) > MAX_HISTORY_ENTRIES:
            self.entries = self.entries[:MAX_HISTORY_ENTRIES]

        self.save()

    # ── Actions ──────────────────────────────────────────────────

    def delete(self, entry_id: str):
        self.entries = [e for e in self.entries if e.id != entry_id]
        self.save()

    def clear_all(self):
        self.entries.clear()
        self.save()

File: macOS/WritingTools/Services/HistoryManager.swift
import Foundation
import Observation
import AppKit

/// Manages the persistent command history log.
@Observable
@MainActor
final class HistoryManager {

    static let shared = HistoryManager()

    // MARK: - State

    /// All entries, sorted newest-first.
    private(set) var entries: [HistoryEntry] = []

    static let maxEntries = 100

    private var storageURL: URL? {
        guard let appSupport = FileManager.default.urls(
            for: .applicationSupportDirectory, in: .userDomainMask
        ).first else { return nil }
        let dir = appSupport.appendingPathComponent("WritingTools", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("command_history.json")
    }

    // MARK: - Init

    private init() {
        loadFromDisk()
    }

    // MARK: - Recording

    /// Records a completed command execution. No-op when history is disabled.
    func record(
        commandName: String,
        commandId: UUID?,
        inputText: String,
        outputText: String,
        modelName: String?,
        sourceApp: NSRunningApplication?,
        matchedProfileName: String?,
        writingCoachPreset: WritingCoachPreset? = nil
    ) {
        guard AppSettings.shared.isHistoryEnabled else { return }
        guard !inputText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }

        let entry = HistoryEntry(
            commandName: commandName,
            commandId: commandId,
            inputText: inputText,
            outputText: outputText,
            modelName: modelName,
            sourceAppName: sourceApp?.localizedName,
            sourceAppBundleId: sourceApp?.bundleIdentifier,
            matchedProfileName: matchedProfileName,
            writingCoachPreset: writingCoachPreset
        )

        entries.insert(entry, at: 0)

        // Auto-prune to max capacity
        if entries.count > Self.maxEntries {
            entries = Array(entries.prefix(Self.maxEntries))
        }

        saveToDisk()
    }

    // MARK: - Deletion

    func delete(_ entry: HistoryEntry) {
        entries.removeAll { $0.id == entry.id }
        saveToDisk()
    }

    func clearAll() {
        entries.removeAll()
        saveToDisk()
    }

    // MARK: - Re-run

    /// Re-runs the command from a history entry using the current prompt text.
    /// Returns false if the command no longer exists in CommandManager.
    @discardableResult
    func rerun(_ entry: HistoryEntry) -> Bool {
        let appState = AppState.shared

        // Find the command by saved ID
        guard let commandId = entry.commandId,
              let command = appState.commandManager.commands.first(where: { $0.id == commandId }) else {
            return false
        }

        // Restore selected text and trigger the command
        if command.responsePresentation == .writingCoach {
            AppSettings.shared.writingCoachPreset = entry.writingCoachPreset ?? .general
        }
        appState.selectedText = entry.inputText
        appState.selectedAttributedText = nil
        appState.processCommand(command)
        return true
    }

    // MARK: - Persistence

    private func loadFromDisk() {
        guard let url = storageURL,
              FileManager.default.fileExists(atPath: url.path),
              let data = try? Data(contentsOf: url) else {
            return
        }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        if let decoded = try? decoder.decode([HistoryEntry].self, from: data) {
            entries = decoded
        }
    }

    private func saveToDisk() {
        guard let url = storageURL else { return }
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        if let data = try? encoder.encode(entries) {
            try? data.write(to: url, options: .atomic)
        }
    }
}
